anydex/wallet/ethereum/eth_database.py

```diff
@@ -40,16 +40,127 @@ class Transaction(Base):
     value = Column(Integer)
     date_time = Column(DateTime, default=datetime.utcnow())
     is_pending = Column(Boolean, default=False)
+    token_identifier = Column(String)  # Used to differentiate between the different tokens.
 
     def __eq__(self, other):
         if not isinstance(other, Transaction):
             raise NotImplementedError(f'cannot compare equality between{self} and {other}')
         return self.hash == other.hash
 
+    def __hash__(self):
+        return hash(self.hash)
+
 
 def initialize_db(db_path):
     engine = create_engine(f'sqlite:///{db_path}', echo=False)
     Base.metadata.create_all(engine)
     session_maker = sessionmaker(bind=engine)
     session = session_maker()
     return session
+
+
+class EthereumDb:
+    """
+    Wrapper around the Ethereum database.
+    """
+
+    def __init__(self, db_path):
+        """
+        :param db_path: full path (including db file)
+        """
+        self.session = initialize_db(db_path)
+
+    def get_wallet_private_key(self, wallet_name):
+        """
+        Retrieve the private key of the wallet with the given name.
+
+        :param wallet_name: name of wallet to find key of
+        :return: private key if wallet exists else None
+        """
+        row = self.session.query(Key).filter(Key.name == wallet_name).first()
+        return row.private_key if row else None
+
+    def add_key(self, name, private_key, address):
+        self.session.add(Key(name=name, private_key=private_key, address=address))
+        self.session.commit()
+
+    def update_database(self, transactions, token):
+        """
+        Update transactions in the database.
+        Pending transactions that have been confirmed will be updated to have a block number and will no longer be
+        pending.
+        Other transactions that are not in the database will be added.
+
+        :param token: token identifier, should be ETH for Ethereum
+        :param transactions: list of transactions
+        """
+        # Use subquery instead??
+        pending_transactions = self.session.query(Transaction).filter(Transaction.is_pending.is_(True)).all()
+        confirmed_transactions = self.session.query(Transaction).filter(Transaction.is_pending.is_(False)).all()
+
+        for transaction in transactions:
```

Comment:
This for loop might be made more efficient if the transactions are selected from the pending transactions using set operations, and then the entire collection is passed to the query. It's unlikely to make a significant difference, but it might be a good refactor option.